### PR TITLE
Scale based rendering with expression compiler

### PR DESCRIPTION
**Pull request by m-kuhn**
This introduces expression compiling for the rule based renderer.

On an example postgis based project with scale based rendering it improved performance by a factor of 3 on small scales. This can vary to a large degree since there are many factors influencing the delay (network latency, server speed, indexes, number of filtered features).

The main change is, that setting a filterRect on a QgsFeatureRequest will no longer be mutually exclusive but combined with other filters (filterFid, filterFids, filterExpression). This can be considered an API break but has been judged an improvement in discussions at the last HF.

~~This is not yet ready for merge as there is still the need to implement the new behavior (non-mutually exclusive treatment of filters) on other providers.~~

### CODE REVIEW

File: src/core/symbology-ng/qgsrendererv2.h [hunk 1]
```diff
@@ -61,6 +61,18 @@ class CORE_EXPORT QgsSymbolV2LevelItem
     int mLayer;
 };
 
+class CORE_EXPORT QgsRenderOptions
```
Comment (nyalldawson) [inline src/core/symbology-ng/qgsrendererv2.h]: Could do with some doxy comments here :)

**Comment by nyalldawson** [comment]
@m-kuhn I'm a bit worried about changing the signature of startRender... won't that potentially break a lot of plugins/PyQgis code which implements custom renderers?

**Comment by m-kuhn** [comment]
Is a change in the return type of a signature detected by python? I think (need to verify) it will just return a default constructed object in this case.

**Comment by wonder-sk** [comment]
Hi @m-kuhn what do you think of having a separate renderer method to return the possible filter expression instead of returning QgsRenderOptions? Most of the renderers do not make use of it anyway, so it would be good to have it as optionally implemented method (and avoiding the new QgsRenderOptions class altogether).

In rule-based renderer:
- the new Rule::startRender() override is largely copy-pasted from the original startRender()... could we get rid of this duplicity?
- is the special "ELSE" rule handled correctly?

The additions looks interesting. I am wondering if complex rules may not actually make the rendering slower - due to the fact that for each feature the expressions will be evaluated twice... would it be possible to compare speed before/after with a shapefile and ~50 rules?

**Comment by m-kuhn** [comment]
@wonder-sk a separate method would be fine as well.
I was thinking that in the future there could be more information returned by the renderer (e.g. order by) which could also be added to QgsRenderOptions. Or we can add another method that returns the order by - or one that modifies a QgsFeatureRequest but this last option would be a bit of a blackbox.

Remove duplicity: Done
Else rule: Done (I hope I got it right)

Check performed on the test data alaska regions, filters with id=1 - id=50: about 2ms of 45ms (5%) performance impact.
We could add a flag "only filter if server side" to QgsFeatureRequest, what do you think about this approach?

File: src/core/symbology-ng/qgsrendererv2.h (hunk 1 above)
Comment (m-kuhn) [inline src/core/symbology-ng/qgsrendererv2.h]: Will do as soon as we know what the final name will be - and if it even stays :)

**Comment by nyalldawson** [comment]
@m-kuhn I'm with Martin in that I think this should be a separate method rather than a class returned by startRender. I'm mostly concerned about future proofing it - it's quite possible that at a later stage we'll need access to the filter outside of when startRender is called. This approach would also avoid any potential API breaks. Was there anything else you were planning on adding to QgsRenderOptions which might change this situation?

**Comment by m-kuhn** [comment]
Changed to a spearate method

~~`QgsRenderOptions startRender();`~~

`QString filter();`

I think the only thing left is the double-evaluation of the expression on providers without native expression support.

**Pull request closed without merging**